src/utils/mod.rs

```diff
@@ -1010,3 +1011,164 @@ pub fn create_unrestricted_signing_rsa_public(
         .build() // should not fail as we control the params
 }
+
+#[derive(Hash, Eq, PartialEq, Copy, Clone, Debug)]
+pub enum PcrSlot {
+    Slot0 = 0,
+    Slot1,
+    Slot2,
+    Slot3,
+    Slot4,
+    Slot5,
+    Slot6,
+    Slot7,
+    Slot8,
+    Slot9,
+    Slot10,
+    Slot11,
+    Slot12,
+    Slot13,
+    Slot14,
+    Slot15,
+    Slot16,
+    Slot17,
+    Slot18,
+    Slot19,
+    Slot20,
+    Slot21,
+    Slot22,
+    Slot23,
+}
+
+impl TryFrom<u32> for PcrSlot {
+    type Error = Error;
+    fn try_from(pcr_slot_number: u32) -> Result<Self> {
+        match pcr_slot_number {
+            0 => Ok(PcrSlot::Slot0),
+            1 => Ok(PcrSlot::Slot1),
+            2 => Ok(PcrSlot::Slot2),
+            3 => Ok(PcrSlot::Slot3),
+            4 => Ok(PcrSlot::Slot4),
+            5 => Ok(PcrSlot::Slot5),
+            6 => Ok(PcrSlot::Slot6),
+            7 => Ok(PcrSlot::Slot7),
+            8 => Ok(PcrSlot::Slot8),
+            9 => Ok(PcrSlot::Slot9),
+            10 => Ok(PcrSlot::Slot10),
+            11 => Ok(PcrSlot::Slot11),
+            12 => Ok(PcrSlot::Slot12),
+            13 => Ok(PcrSlot::Slot13),
+            14 => Ok(PcrSlot::Slot14),
+            15 => Ok(PcrSlot::Slot15),
+            16 => Ok(PcrSlot::Slot16),
+            17 => Ok(PcrSlot::Slot17),
+            18 => Ok(PcrSlot::Slot18),
+            19 => Ok(PcrSlot::Slot19),
+            20 => Ok(PcrSlot::Slot20),
+            21 => Ok(PcrSlot::Slot21),
+            22 => Ok(PcrSlot::Slot22),
+            23 => Ok(PcrSlot::Slot23),
+            _ => Err(Error::local_error(WrapperErrorKind::InvalidParam)),
+        }
+    }
+}
+
+/// TPML_PCR_SELECTION is a structure that can contain up to
+/// 16 TPMS_PCR_SELECTION
+///
+/// The TPMS_PCR_SELECTION is a variable size bitmask
+/// where the position of each bit is the selected index.
+///
+/// The minimum number of octets in a TPMS_PCR_SELECT.sizeOfSelect
+/// is not determined by the number of PCR implemented but by the
+/// number of PCR required by the platform-specific
+/// specification with which the TPM is compliant or by the implementer if
+/// not adhering to a platform-specific specification.
+///
+/// So to make things simple 3 octets are
+/// used which means that size of selections always will be
+/// 3.
+// For example, selecting PCRs 3 and 9 looks like:
+// size(3)  mask     mask     mask     mask
+// 00000011 00000000 00000000 00000001 00000100
+#[derive(Debug, Default, Clone)]
+pub struct PcrSelections {
+    items: HashMap<HashingAlgorithm, HashSet<PcrSlot>>,
+}
+
+impl From<PcrSelections> for TPML_PCR_SELECTION {
+    fn from(pcr_selections: PcrSelections) -> TPML_PCR_SELECTION {
+        let mut ret: TPML_PCR_SELECTION = Default::default();
+        for (hash_algorithm, pcr_slots) in &pcr_selections.items {
+            ret.pcrSelections[ret.count as usize].hash = hash_algorithm.clone().into();
+            ret.pcrSelections[ret.count as usize].sizeofSelect = 3;
```

Comment:
I'm weary about that hardcoded constant; from the spec:

> sizeofSelect indicates  the  number  of  octets  in pcrSelect.  The  allowable  value  for sizeofSelectis determined  by  the  number  of  PCR  required  by  the  applicable  platform-specific  specification  and  the number of PCR implemented in the TPM

[Mobile TPMs](https://trustedcomputinggroup.org/wp-content/uploads/TPM_2.0_Mobile_Common_Profile_v2r31_FINAL.pdf), for example, can contain as little as 8 PCRs. 

I think the safest bet here would be to either: 
* move this code into a method implemented on `PcrSelection` so that you can take this value in as a param
* add another field in `PcrSelections` that holds this value and then you can default to 3 if it's not set in the Builder

Reply:
I think you are right and this is good idea and it has now been fixed.